### ISSUE TITLE
False positive: needless_range_loop

### ISSUE THREAD

**Issue by fuzzbuck** [rust-lang/rust-clippy#11399]
### Summary

Related code:
```rust
for i in 0..message.len() {
        block.data[i] = message[i] as char;
 }
 ```

Clippy begs me to use ```rust
for x in data```
without the 'i' enumerator, however this would make fetching message[i] impossible.

### Lint Name

needless_range_loop

### Reproducer

I tried this code:

```rust
for i in 0..message.len() {
        block.data[i] = message[i] as char;
}
```

I saw this happen:

```
the loop variable `i` is used to index `message` Help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_range_loop Help: consider using an iterator and enumerate()
```

I expected to see this happen:

no complaints from Clippy

### Version

```text
rustc 1.72.0-nightly (8c74a5d27 2023-06-14)
binary: rustc
commit-hash: 8c74a5d27c644a0f7a22bb2fa8dd3ff8257bc220
commit-date: 2023-06-14
host: x86_64-pc-windows-msvc
release: 1.72.0-nightly
LLVM version: 16.0.5
```

**Comment by y21** [comment]
You still get the `i` with clippy's suggestion though
```rs
help: consider using an iterator and enumerate()
   |
25 |     for (i, <item>) in message.iter().enumerate() {
   |         ~~~~~~~~~~~    ~~~~~~~~~~~~~~~~~~~~~~~~~~

```

**Comment by fuzzbuck** [comment]
> You still get the `i` with clippy's suggestion though
> ```rs
> help: consider using an iterator and enumerate()
>    |
> 25 |     for (i, <item>) in message.iter().enumerate() {
>    |         ~~~~~~~~~~~    ~~~~~~~~~~~~~~~~~~~~~~~~~~
> 
> ```

I would imagine such suggestion is in place to prevent needless indexing variables in for loops
e.
```
let i = 0;
for value in some_vec {
    println!("{} = {}");
    i += 1;
}```
In such case the suggestion is valid, however for the problem provided in the issue both solutions are valid, and from my (biased) opinion the former syntax is more readable. (in contrast to enumerate)
